### PR TITLE
BUG: stats: Implementing boost's hypergeometric distribution to correct numerical errors

### PR DESCRIPTION
**Pull request by V0lantis**
Follows my comment #13367. I added the tests that @WarrenWeckesser did on the PR. Instead of reimplementing hypergeom, I am using boost's c++ hypergeometric distrubution implementation.

Two avantages : Improvement of precision and improvement of performance (on local testing, I noticed a 10 times improvement while timing the old and the new version). 

I also added a Boschloo's pvalue computation improvement in 4d875dd80645147b256d9998533a55963e6c82b0, to correct numerical errors
@mdhaber, what do you think about it ? If all tests are passing, I'll try to remove `pytest.mark.xfail_on_32bit` to see if it passes. I am quite confident it will. 

@mckib2, could you check if I am doing everything right?

### CODE REVIEW

File: scipy/stats/_hypotests.py [hunk 1]
```diff
@@ -1144,6 +1144,7 @@ def boschloo_exact(table, alternative="two-sided", n=32):
 
     fisher_stat = pvalues[table[0, 0], table[0, 1]]
     index_arr = pvalues <= fisher_stat
+    index_arr += np.isclose(pvalues - fisher_stat, 0)  # Because the hypergeometric has some slight numerical errors
```
Comment (mdhaber) [inline scipy/stats/_hypotests.py:1147]: I would prefer to put this in a separate PR, as it will take some discussion. For instance, currently it appears that this could double-count some `pvalues` that are less than `fisher_stat` and also close to `fisher_stat`.  Also, can you determine a theoretical bound on how close is close enough (i.e. definitely smaller than the smallest legitimate difference between pvalues)?

File: scipy/stats/_discrete_distns.py [hunk 2]
```diff
@@ -504,44 +504,33 @@ def _logpmf(self, k, M, n, N):
...
-        return mu, var, g1, g2
+        g2 = M * (M + 1) - 6. * N * (M - N) - 6. * n * m
+        g2 *= (M - 1) * M * M
+        g2 += 6. * n * N * (M - N) * m * (5. * M - 6)
+        g2 /= n * N * (M - N) * m * (M - 2.) * (M - 3.)
+        return (
+            _boost._hypergeom_mean(n, N, M),
+            _boost._hypergeom_variance(n, N, M),
+            _boost._hypergeom_skewness(n, N, M),
+            g2,  # Boost kurtosis_excess doesn't return the same as the value computed here.
```
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:524]: Does it have to do with the difference between kurtosis and excess/Fisher kurtosis, or does it look wrong?

Comment (mckib2) [reply scipy/stats/_discrete_distns.py:524]: If I understand the [wikipedia](The excess kurtosis is defined as kurtosis minus 3. There are 3 distinct regimes as described below. ) correctly, you should just be able to add 3 to get kurtosis from excess kurtosis?

**Comment by mckib2** [comment]
@V0lantis Usage of the stats Boost wrapper generation looks good to me!  Interested to hear about the discrepancy in kurtosis, but otherwise this is looking really good to me

File: scipy/stats/_discrete_distns.py (hunk 2 above)
Comment (V0lantis) [reply scipy/stats/_discrete_distns.py:524]: I wasn't able to get the same values as expected. Ill dig into it tonight

**Comment by V0lantis** [comment]
Ok, so here is were my research leaded. The kurtosis's formula for the hypergeometric distribution from wikipedia is : 
Capture d’écran 2021-05-30 à 19 20 08

It is a little bit messy. 

I found the [boost's implementation](https://github.com/boostorg/math/blob/45a2cbe789e1b81d310024d58c9616cc434a63d3/include/boost/math/distributions/hypergeometric.hpp#L269) : 

```c++
inline RealType kurtosis_excess(const hypergeometric_distribution<RealType, Policy>& dist)
   {
      RealType r = static_cast<RealType>(dist.defective());
      RealType n = static_cast<RealType>(dist.sample_count());
      RealType N = static_cast<RealType>(dist.total());
      RealType t1 = N * N * (N - 1) / (r * (N - 2) * (N - 3) * (N - r));
      RealType t2 = (N * (N + 1) - 6 * N * (N - r)) / (n * (N - n))
         + 3 * r * (N - r) * (N + 6) / (N * N) - 6;
      return t1 * t2;
   } // RealType kurtosis_excess(const hypergeometric_distribution<RealType, Policy>& dist)
```

As a reminder, here is the old's Scipy one (I simplified it to stay concise) :
```python
def _stats(self, M, n, N):
    # tot, good, sample_size = M, n, N
    # "wikipedia".replace('N', 'M').replace('n', 'N').replace('K', 'n')
    ...
    g2 = M*(M+1) - 6.*N*(M-N) - 6.*n*m
    g2 *= (M-1)*M*M
    g2 += 6.*n*N*(M-N)*m*(5.*M-6)
    g2 /= n * N * (M-N) * m * (M-2.) * (M-3.)
    return mu, var, g1, g2
``` 

Here is Boost computation formula : 

Capture d’écran 2021-05-30 à 19 47 21

Boost after changing r to K (to stick to wikipedia notation)

Capture d’écran 2021-05-30 à 19 49 49

Then I try to factorise wikipedia formulas to find boost one; by : 
Capture d’écran 2021-05-30 à 19 50 42

Capture d’écran 2021-05-30 à 19 51 08


There are a lot of difference and therefore I am wondering if there might be an error in boost implementation. Therefore, could one of you check with my calculations to see if I didn't do any mistakes? I am usually quite prone to some while doing calculations.

**Comment by mckib2** [comment]
@V0lantis It appears Boost's kurtosis formula is non-standard or incorrect.  I have submitted an upstream bug but it might be a while until we can get that fix, so keeping the Python `g2` kurtosis calculation seems to be the right thing to do for now

**Comment by pabristow** [comment]
Looks suspiciously like a bug in Boost implementation.  I'm on the case :-(

**Comment by mckib2** [comment]
> @mckib2 do you think we could integrate boost's fix in the future ? I saw that Scipy is currently using Boost v1.74, while v1.76 is already out. Is there any reason for that ?

SciPy is actually pointed at [this commit](https://github.com/scipy/boost-headers-only/tree/2110ce20e59917b85b12059b3ddd8b133549f662) of the boost-headers-only repo which is at version 1.75.  As explained in [this comment](https://github.com/scipy/scipy/pull/13328#issuecomment-832403900) on the original Boost PR, upgrading past 1.7.5 would require bumping the minimum GCC version to at least 7.1.0 which is a while off for us.

Another option might be modifying the [boost-headers-only scripts](https://github.com/scipy/boost-headers-only/blob/master/make_headers.py) to apply SciPy specific patches or use a more recent version Boost.Math when generating the headers.  I haven't tried doing that, so I don't know how difficult it would be

**Comment by V0lantis** [comment]
I remove 
pytest.param(
                [[0, 1], [3, 2]], (0.5, 0.6875),
                marks=pytest.mark.xfail_on_32bit(
                    "This test is sensitive to machine epsilon level",
                    reason="numerical error"),
            ),
```
in `less` and `two-sided` Boschloo tests, and it is working just fine, which was the primary reason for this PR (and the implementation of the boost Hypergeometric distribution).

Failure seems unrelated to this PR

File: scipy/stats/_hypotests.py [hunk 3]
```diff
@@ -1144,6 +1144,10 @@ def boschloo_exact(table, alternative="two-sided", n=32):
 
     fisher_stat = pvalues[table[0, 0], table[0, 1]]
     index_arr = pvalues <= fisher_stat
+    index_arr = np.logical_or(
+        index_arr,
+        np.isclose(pvalues - fisher_stat, 0)
+    )  # Because the hypergeometric has some slight numerical errors
```
Comment (mdhaber) [inline scipy/stats/_hypotests.py:1150]: I thought I'd link [the conversation](https://github.com/scipy/scipy/pull/14153/files/4d875dd80645147b256d9998533a55963e6c82b0#diff-6ba7fcf5a871c6742dd082e5fc1801ef3ce2126456b052e9d0a126e70aab5f99) to the new relevant line.

I understand that you included this so that the tests would pass. In the interest of decoupling the clear improvement to the hypergeometric distribution from this debatable change to `boschloo_exact`, I would suggest instead xfailing the problematic test case and opening a separate issue/PR for this change. 

If you really want to include this change to `boschloo_exact` here, please address the comment:
> Can you determine a theoretical bound on how close is close enough (i.e. definitely smaller than a legitimate difference between pvalues)? I think you can - it's something like the `hypergeom.pmf` near `fisher_stat`.

This is important, because `np.isclose` with default tolerances will not do the right thing when the difference between legitimately distinct elements of the distribution are small. 

I'll support the `boschloo_exact` changes, whether they're left in this PR or moved to a separate PR, but I will let @WarrenWeckesser and @mckib2 review the rest of this.

**Comment by pabristow** [comment]
After some study, and much confusion caused by different variable name and argument ordering, and the excess, I conclude that Boost.Math kurtosis in most probably wrong.

I am unable to trace the origin of the formula used by Boost.Math, and the Boost.Math test value 2.5155486690782807 does not agree with Wolfram (Wolfram always calculates the kurtosis - even when the excess is requested, and not the kurtosis_excess, hence -3 is needed)
``
N[ -3 + kurtosis[hypergeometricdistribution(200,50,500)], 55] 
-0.03082596426394152509883037627819572499701466271973653613
``
nor with skipy
``
import numpy as np
from scipy.stats import hypergeom
[M, n, N] =[500, 50, 200]
d = hypergeom(M, n, N)
s = d.stats('msvk')
(array(20.), array(10.82164329), array(0.04883307), array(-0.03082596))
mean = 20, variance = 10.82164329, skewness = 0.04883307 kurtosis = -0.03082596
``
nor with my implementation of Wikipedia formula or my implementation using a plain-text single line formula from Wolfram converted to a C++ expression (without any concern for overflow and assuming that the compile will optimise).

In mitigation, I plead this quotation from Wolfram  https://mathworld.wolfram.com/HypergeometricDistribution.html

   **"the kurtosis excess is given by a complicated expression".** ;-)

Subject to review, either of these two could be used to correct Boost's kurtosis_excess for the next Boost release.

Meanwhile, your 'fudge' using g2 looks correct as a work-around to get the test to 'pass'.

Or I could provide a small patch to change the formula in 
``template <class RealType, class Policy>
   inline RealType kurtosis_excess(const hypergeometric_distribution<RealType, Policy>& dist)
   {
      RealType r = static_cast<RealType>(dist.defective());
      RealType n = static_cast<RealType>(dist.sample_count());
      RealType N = static_cast<RealType>(dist.total());
      RealType t1 = N * N * (N - 1) / (r * (N - 2) * (N - 3) * (N - r));
      RealType t2 = (N * (N + 1) - 6 * N * (N - r)) / (n * (N - n))
         + 3 * r * (N - r) * (N + 6) / (N * N) - 6;
      return t1 * t2;
   } 
``
``
      RealType m = static_cast<RealType>(dist.defective()); // Failures or success. 50
      RealType n = static_cast<RealType>(dist.sample_count()); // draws, trials. = 200
      RealType N = static_cast<RealType>(dist.total()); // total population from which n draws or trials are made.  500
      RealType N2 = N * N; // N^2
      RealType n2 = n * n; // n^2
      RealType result = ((N-1)*N2*((3*m*(N-m)*(n2*(-N)+(n-2)*N2+6*n*(N-n)))/N2-6*n*(N-n)+N*(N+1)))/(m*n*(N-3)*(N-2)*(N-m)*(N-n));
      return result -3;  // Excess kurtosis.
``
These agree with Wolfram to 128-bit precision using Boost.Multiprecision.

All other moments appear correct.

HTH and tell if I can help more.

Aside: Boost now requires C++11, but is likely to move up closer to the most recent C++ Standard.
This has been done to reduce the massive Macro complexity to deal with 'dodgy' compilers and to replace Boost versions of functions that have been adopted and are now in std::.  A Boost.Math Standalone version with almost no outside dependencies (so avoiding the need to full in all of Boost) is under test, but will require a **much** more recent compiler than GCC 7.
GCC 9 has proved problematic and GCC 11 is recommended.  Despite the pain, moving more rapidly to the most recent C++ compiler versions should be a long term objective.

**Comment by mckib2** [comment]
Thanks for looking into this @pabristow !

This PR looks good to me, I don't know if there's a possibility of getting into the next release because we've already branched.  I don't quite track the `boschloo_exact` changes, but if @WarrenWeckesser or @mdhaber give the thumbs up I'll plan on merging over the weekend to give people a chance to look at it

**Comment by mckib2** [comment]
> HTH and tell if I can help more.

Again, thanks for looking at this @pabristow , I had suspected it was a bug, but also thought there might have been something else I didn't understand.

> Aside: Boost now requires C++11, but is likely to move up closer to the most recent C++ Standard.
This has been done to reduce the massive Macro complexity to deal with 'dodgy' compilers and to replace Boost versions of functions that have been adopted and are now in std::. A Boost.Math Standalone version with almost no outside dependencies (so avoiding the need to full in all of Boost) is under test, but will require a much more recent compiler than GCC 7.
GCC 9 has proved problematic and GCC 11 is recommended. Despite the pain, moving more rapidly to the most recent C++ compiler versions should be a long term objective.

There's a short discussion of C++ standard support in the [toolchain roadmap](https://docs.scipy.org/doc/scipy/reference/toolchain.html#c-and-c-language-standards) that you might be interested in.  We tend to move pretty cautiously and slowly when it comes to deprecating compilers and moving to require newer C++ standards.  There was talk last year about moving minimum GCC from v4.8 to v5, but I got busy with other things and I think that has stalled out.  It is absolutely a long term objective, but history has shown us that users still need to build SciPy using older compilers that are sometimes not feasibly able to be upgraded (I encounter machines like that everyday at my day job).  From what I can tell, Python's binary wheel requirements drive our deprecation cycles here, so we will likely never be targeting bleeding-edge compilers

File: scipy/stats/_hypotests.py (hunk 3 above)
Comment (V0lantis) [reply scipy/stats/_hypotests.py]: I removed Boschloo Updates and add an `xfail` for the test which is sensitive to epsilon machine. I am going to submit an other PR to correct this. 
If this is ok for you this PR is ready to be merged according to @pabristow and @mckib2

**Comment by V0lantis** [comment]
Thank you @mckib2 and @pabristow for your comments and research :-)

File: scipy/stats/tests/test_hypotests.py [hunk 4]
```diff
@@ -1019,11 +1019,11 @@ class TestBoschlooExact:
             ([[10, 5], [10, 1]], (0.1652174, 0.08565611)),
             ([[5, 0], [1, 4]], (1, 1)),
             pytest.param(
-                [[0, 1], [3, 2]], (0.5, 0.34375),
-                marks=pytest.mark.xfail_on_32bit(
-                    "This test is sensitive to machine epsilon level",
-                    reason="numerical error"),
-            ),
+                 [[0, 1], [3, 2]], (0.5, 0.34375),
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_hypotests.py:1022]: Why did you increase the indentation (here and below) from 4 to 5 spaces?  Generally 4 is preferred.

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
Looks good; I have just a couple formatting changes requested.

File: scipy/stats/tests/test_hypotests.py (hunk 4 above)
Comment (V0lantis) [reply scipy/stats/tests/test_hypotests.py]: Ok, it slipped from my eyes when I wrote this. Corrected it in 920bbf3c55f5da59dab237c42ba23e4c112b20d2

File: scipy/stats/_discrete_distns.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (V0lantis) [reply scipy/stats/_discrete_distns.py]: Corrected in a52c759e5d11f337d3547e7c8c26242105d62dbf

**Review by V0lantis — COMMENTED** [review]
Thank you @WarrenWeckesser for this checkup :-)

**Comment by V0lantis** [comment]
@WarrenWeckesser @mdhaber LGTM and ready to be merged :-)

**Comment by mdhaber** [comment]
@Volantis i mentioned [here](https://github.com/scipy/scipy/pull/14153#discussion_r643515770) that i would review the Boschloo PR, but I'd prefer to let others review this one. Thanks for submitting it, though; it looks like it will be a big improvement.

**Comment by mckib2** [comment]
Looks like comments were addressed and LGTM, so I'll merge.  Thanks @V0lantis !

**Comment by pabristow** [comment]
Boost error in cmputing kurtosis has been corrected in Boost develop branch

https://github.com/boostorg/math/commit/d35f821eefe3d84bf28bb38f74b4d2faf4aedc99 

and will appear in the next release Boost 1.77 fairly soon.

**Comment by V0lantis** [comment]
> Boost error in cmputing kurtosis has been corrected in Boost develop branch
> 
> 
> 
> https://github.com/boostorg/math/commit/d35f821eefe3d84bf28bb38f74b4d2faf4aedc99 
> 
> 
> 
> and will appear in the next release Boost 1.77 fairly soon.

Great, thank you so much @pabristow !

**Comment by NAThompson** [comment]
@V0lantis : We have created a [standalone version](https://github.com/boostorg/math/releases/tag/v1.76-standalone-beta) of boost.math, and we'll shortly be creating a "blessed" release.

This can perhaps help you guys in that 1) vendorizing boost.math might not be necessary, as it could be consumed as a smaller git submodule than all of boost, and 2) the cadence can increase, as we can just cut a release any time we find a bug like this.

Not fully aware of how y'all are consuming boost, but hopefully this at least provides an option which is more attractive than the current workflow.

**Comment by mckib2** [comment]
One of the main blockers to moving to newer version of Boost is compiler support.  I believe need GCC 4.8 and Boost > v1.75 doesn't build with this config.  Do you know if your standalone version supports GCC version that old?

**Comment by NAThompson** [comment]
Yeah, 4.8 is tough. We just decided to remove it from CI because the workarounds got too painful. But we didn't actively *remove* gcc 4.8 support.

So depending on what you use, you *might* get away with it.
